spec/ics-4-channel-and-packet-semantics/README.md

```diff
@@ -0,0 +1,384 @@
+
+### Motivation
+
+The interblockchain communication protocol uses a cross-chain message passing model which makes no assumptions about network synchrony. IBC *packets* are relayed from one blockchain to the other by external relayer processes. Chain `A` and chain `B` confirm new blocks independently, and packets from one chain to the other may be delayed, censored, or re-ordered arbitrarily. Packets are public and can be read from a blockchain by any relayer and submitted to any other blockchain. In order to provide the desired ordering, exactly-once delivery, and module permissioning semantics to the application layer, the interblockchain communication protocol must implement an abstraction to enforce these semantics — channels are this abstraction.
+
+### Definitions
+
+`Connection` is as defined in [ICS 3](../ics-3-connection-semantics).
+
+`Commitment`, `CommitmentProof`, and `CommitmentRoot` are as defined in [ICS 23](../ics-23-vector-commitments).
+
+`Identifier`, `get`, `set`, `delete`, and module-system related primitives are as defined in [ICS 24](../ics-24-host-requirements).
+
+A *channel* is a pipeline for exactly-once packet delivery between specific modules on separate blockchains, which has at least one send and one receive end.
+
+A *bidirectional* channel is a channel where packets can flow in both directions: from `A` to `B` and from `B` to `A`.
+
+A *unidirectional* channel is a channel where packets can only flow in one direction: from `A` to `B`.
+
+An *ordered* channel is a channel where packets are delivered exactly in the order which they were sent.
+
+An *unordered* channel is a channel where packets can be delivered in any order, which may differ from the order in which they were sent.
+
+Directionality and ordering are independent, so one can speak of a bidirectional unordered channel, a unidirectional ordered channel, etc.
+
+All channels provide exactly-once packet delivery, meaning that a packet sent on one end of a channel is delivered no more than once, eventually, to the other end.
+
+This specification only concerns itself with *bidirectional ordered* channels. *Unidirectional* and *unordered* channels use almost exactly the same protocol and will be outlined in a future ICS.
+
+Channels have a *state*:
+
+```typescript
+enum ChannelState {
+  INIT,
+  TRYOPEN,
+  OPEN,
+  TRYCLOSE,
+  CLOSED,
+}
+```
+
+An *end* of a channel is a data structure on one chain storing channel metadata:
+
+```typescript
+interface ChannelEnd {
+  state: ChannelState
+  counterpartyChannelIdentifier: Identifier
+  moduleIdentifier: Identifier
+  counterpartyModuleIdentifier: Identifier
+  nextSequenceSend: uint64
+  nextSequenceRecv: uint64
+}
+```
+
+An IBC *packet* is a particular datagram, defined as follows:
+
+```typescript
+interface Packet {
+  sequence: uint64
+  timeoutHeight: uint64
+  sourceConnection: Identifier
+  sourceChannel: Identifier
+  destConnection: Identifier
+  destChannel: Identifier
+  data: bytes
+}
+```
```

Comment:
Does this describe datagrams for channel establishment also? If not, can we move this in a "Packet handling" section together with send, receive, timeout?

Reply:
So moved.